libswift: implement `ReleaseDevirtualizer` in Swift

MaxDesiatov opened the pull request:
Implements one more SILOptimizer pass in Swift.

MaxDesiatov committed:
libswift: bridge `RCIdentityAnalysis` code

MaxDesiatov committed:
libswift: clean up `Instruction.swift` formatting

MaxDesiatov committed:
libswift: clean up `SIL/Utils.swift` formatting

MaxDesiatov committed:
libswift: clean up `SIL/Type.swift` formatting

MaxDesiatov committed:
libswift: clean up `SIL/Function.swift` formatting

MaxDesiatov committed:
libswift: add `SubstitutionMap` to Swift code

MaxDesiatov committed:
libswift: bridge more functions from `SILBuilder`

gottesmm commented:
Can you be sure whenever you are porting functions to also add an additional test line to the test that tests the pass. I think we still should test both the legacy/swift pass for now. For instance, you could add an extra test line to this file test/SILOptimizer/devirt_release.sil that way we make sure both passes are correct.

MaxDesiatov commented:
> Can you be sure whenever you are porting functions to also add an additional test line to the test that tests the pass. I think we still should test both the legacy/swift pass for now. For instance, you could add an extra test line to this file test/SILOptimizer/devirt_release.sil that way we make sure both passes are correct.

What extra line should be added there? Is there a `sil-opt` flag that allows switching between legacy and new Swift passes?

MaxDesiatov commented:
> I think we still should test both the legacy/swift pass for now.

Also, this will make Windows tests fail, since Swift passes aren't built for Windows yet.

MaxDesiatov commented:
@eeckstein the Swift bridge to `RCIdentityAnalysis` is going to be removed as soon as I'm sure my use-def walk is good enough, but I wanted to get some preliminary feedback on the overall approach. Thanks!

eeckstein commented on SwiftCompilerSources/Sources/CMakeLists.txt:
```diff
@@ -9,5 +9,6 @@
 if(SWIFT_ENABLE_EXPERIMENTAL_STRING_PROCESSING)
   add_subdirectory(ExperimentalRegex)
 endif()
+add_subdirectory(AST)
```
Please put SubstitutionMap directly into SIL. 
It's like SILType, which is a wrapper around the AST Type. Except that in C++ we don't have a SIL wrapper around SubstitutionMap, but that doesn't matter for the swift SubstitutionMap.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/Analysis/CMakeLists.txt:
```diff
@@ -1,11 +1,12 @@
   AliasAnalysis.swift
-  CalleeAnalysis.swift)
+  CalleeAnalysis.swift
+  RCIdentityAnalysis.swift)
```
As you said, please remove that when you have your own use-def walk.
Please put your own utility function for that directly into ReleaseDevirtualizer.swift.
As we discussed offline, in future we can use the new EscapeInfo utility to do a better analysis.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,140 @@
+  _ functionInfo: RCIdentityFunctionInfo
+) {
+  let allocRefInstruction = deallocStackRef.allocRef
+  let rcRoot = functionInfo.getRCIdentityRoot(release.operands[0].value)
```
Please rename `reroot` to something like `rootObject`

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,140 @@
+
+  precondition(rcRoot === root)
+
+  guard rcRoot === allocRefInstruction else {
```
No need for a `guard`. You can use an `if`.
Also, no need for `===`. `Value` has a `==` operator.

MaxDesiatov replied:
`allocRefInstruction` here is of `AllocRefInst` type, while `root` is of `Value`. For some reason `==` does not automatically cast `AllocRefInst` to `Value`, would you prefer me to cast explicitly, or keep `===` as it is?

MaxDesiatov replied:
Also, `Value` isn't `Equatable` right now, should the former inherit from the latter? Otherwise `!=` isn't currently defined on `Value` arguments to convert this from `guard` to `if` in a nice way.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,140 @@
+    return
+  }
+
+  guard rcRoot === release.operands[0].value else {
```
I guess this is just some testing code and you'll remove it, right?

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/PassManager/PassUtils.swift:
```diff
@@ -84,6 +89,14 @@ struct PassContext {
     RefCountingInst_setIsAtomic(instruction.bridged, isAtomic)
   }
+
+  func getDealloc(for type: Type) -> Function? {
```
Please rename this function to `getDestructor(ofClass: Type)`.
We use the term "Destructor" also in the AST, e.g. `DestructorDecl`.
And as we don't have separate swift types for the different AST types (we just pack everything into `Type`), it should be clear from the label that the argument should be a class type.

eeckstein commented on SwiftCompilerSources/Sources/SIL/Builder.swift:
```diff
@@ -67,4 +68,42 @@ public struct Builder {
+    _ substitutionMap: SubstitutionMap,
+    arguments: [Value]
+  ) -> FunctionRefInst {
+    notifyInstructionsChanged()
```
You also need to `notifyCallsChanged()` here.

eeckstein commented on SwiftCompilerSources/Sources/SIL/Builder.swift:
```diff
@@ -67,4 +68,42 @@ public struct Builder {
+    function: Value,
+    _ substitutionMap: SubstitutionMap,
+    arguments: [Value]
+  ) -> FunctionRefInst {
```
`FunctionRefInst` -> `ApplyInst`

eeckstein commented on SwiftCompilerSources/Sources/SIL/Builder.swift:
```diff
@@ -67,4 +68,42 @@ public struct Builder {
+        substitutionMap.bridged, valuesRef
+      )
+    }
+    return functionRef.getAs(FunctionRefInst.self)
```
`FunctionRefInst` -> `ApplyInst`

eeckstein commented on SwiftCompilerSources/Sources/SIL/Builder.swift:
```diff
@@ -67,4 +68,42 @@ public struct Builder {
+    arguments: [Value]
+  ) -> FunctionRefInst {
+    notifyInstructionsChanged()
+    let functionRef = arguments.withBridgedValues { valuesRef in
```
`functionRef` -> `apply`

eeckstein reviewed (changes requested):
Nice work!
See my comments inline.

eeckstein commented:
> I think we still should test both the legacy/swift pass for now.

@gottesmm Legacy passes are tested with the Windows PR tests. So no need to test them on macOS/Linux (they don't run on these platforms anyway).

gottesmm commented:
@eeckstein ok. My fear is that will be forgotten. Perhaps we add notes into each of both passes that explain this? Otherwise I am scared that people will forget (I probably will with enough time). Maybe write that in a comment in the individual test itself? That would be good enough.

eeckstein committed:
swift SIL: add some SIL type related bridging

* `Function.argumentTypes` and `Function.resultType`
* `Type.isNominal`, `Type.isClass`, `Type.isTuple`, `Type.isStruct` and `Type.isEnum`
* `Type.getFieldIndexOfNominal`
* `Type.getFieldTypeOfNominal`
* `Type.tupleElements`
* `Type.description` for better debugging

# Conflicts:
#	SwiftCompilerSources/Sources/SIL/Type.swift
#	SwiftCompilerSources/Sources/SIL/Utils.swift
#	SwiftCompilerSources/Sources/SIL/Value.swift
#	include/swift/SIL/SILBridging.h
#	lib/SIL/Utils/SILBridging.cpp

MaxDesiatov committed:
libswift: implement `ReleaseDevirtualizer` in Swift

MaxDesiatov commented:
@swift-ci please smoke test

MaxDesiatov commented:
@swift-ci Please smoke benchmark

MaxDesiatov commented on lib/SILOptimizer/PassManager/PassManager.cpp:
```diff
@@ -1243,8 +1244,8 @@ void PassContext_fixStackNesting(BridgedPassContext passContext,
-  return AvailabilityContext::forDeploymentTarget(ctxt).
-           isContainedIn(ctxt.getSwift51Availability());
+  return AvailabilityContext::forDeploymentTarget(ctxt).isContainedIn(
+    ctxt.getSwift51Availability());
```
This change was applied by the formatter.

MaxDesiatov commented on lib/SILOptimizer/Transforms/ReleaseDevirtualizer.cpp:
```diff
@@ -133,15 +133,10 @@ bool ReleaseDevirtualizer::createDeallocCall(SILType AllocType,
     ->getContextSubstitutionMap(M.getSwiftModule(), NTD);
 
-  DeallocType = DeallocType->substGenericArgs(M, AllocSubMap, context);
-
```
Removed dead code here

MaxDesiatov commented:
@eeckstein all changes were addressed, I've verified that use-def walk behaves in the same way as it does in the C++ version. `RCIdentityAnalysis` use was cleaned up in the Swift version and it's no longer bridged.

@gottesmm I've added a comment to the test that clarifies how the legacy and the new Swift pass are tested on Windows and other platforms.

MaxDesiatov commented:
@swift-ci please smoke test

MaxDesiatov commented:
@swift-ci please benchmark

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,160 @@
+    root = newRoot
+  }
+
+  guard root === allocRefInstruction else {
```
* no need to use a guard. An `if root != allocRefInstruction { return }` works as well
* you can use `==` instead of `===`

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,160 @@
+)
+
+/// Devirtualize releases of array buffers.
+private func devirtualizeReleaseOfObject(
```
a better name would be `tryDevirtualizeReleaseOfObject` because it's not guaranteed to be done

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,160 @@
+
+/// Replaces the release-instruction `release` with an explicit call to
+/// the deallocating destructor of `type` for `object`.
+private func createDeallocCall(
```
you can merge this function with `tryDevirtualizeReleaseOfObject`. IMO, `createDeallocCall` is confusing, because it can fail (if the destructor of the type is not found).

Or: move the `guard let dealloc...` into `tryDevirtualizeReleaseOfObject`

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,160 @@
+    return
+  }
+
+  let substitutionMap = context.getContextSubstitutionMap(for: type)
```
This line can be moved down to before the `builder.createApply`

eeckstein commented on SwiftCompilerSources/Sources/SIL/Instruction.swift:
```diff
@@ -98,6 +98,10 @@ public class Instruction : ListNode, CustomStringConvertible, Hashable {
     }
   }
 
+  final public var mayReleaseOrReadRefCount: Bool {
```
I would prefer to have two separate properties: `mayRelease` and `mayReadRefCount`.
`mayReadRefCount` can be implemented in swift (no need to bridge it).

eeckstein commented on SwiftCompilerSources/Sources/SIL/Instruction.swift:
```diff
@@ -356,10 +362,14 @@ final public class GlobalValueInst : GlobalAccessInst {}
 final public class IntegerLiteralInst : SingleValueInstruction {}
 
 final public class TupleInst : SingleValueInstruction {
+  public var uniqueNonTrivialElt: Value? {
```
This is actually a utility and not a property of the instruction. Please move this to `ReleaseDevirtualizer.swift` (I don't expect it to be used by any other pass).
You can implement that in swift, no need to bridge it.

eeckstein commented on SwiftCompilerSources/Sources/SIL/Instruction.swift:
```diff
@@ -356,10 +362,14 @@ final public class GlobalValueInst : GlobalAccessInst {}
 
 final public class TupleExtractInst : SingleValueInstruction, UnaryInstruction {
   public var fieldIndex: Int { TupleExtractInst_fieldIndex(bridged) }
+  public var isEltOnlyNonTrivialElt: Bool { TupleExtractInst_isEltOnlyNonTrivialElt(bridged) }
```
The same here

eeckstein commented on SwiftCompilerSources/Sources/SIL/Instruction.swift:
```diff
@@ -368,10 +378,16 @@ class TupleElementAddrInst : SingleValueInstruction, UnaryInstruction {
 }
 
 final public class StructInst : SingleValueInstruction {
+  public var uniqueNonTrivialFieldValue: Value? {
```
and here

eeckstein commented on SwiftCompilerSources/Sources/SIL/Instruction.swift:
```diff
@@ -368,10 +378,16 @@ class TupleElementAddrInst : SingleValueInstruction, UnaryInstruction {
 
 final public class StructExtractInst : SingleValueInstruction, UnaryInstruction {
   public var fieldIndex: Int { StructExtractInst_fieldIndex(bridged) }
+  public var isFieldOnlyNonTrivialField: Bool { 
```
and here

MaxDesiatov replied:
This one needs to access some AST types, like `TypeExpansionContext` and `VarDecl`. Should I bridge those types, or keep whole `StructExtractInst_isFieldOnlyNonTrivialField` bridged?

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,160 @@
+  }
+)
+
+/// Devirtualize releases of array buffers.
```
This comment is outdated. The optimization not specific to array buffers.
Maybe something like `Tries to de-virtualize the final release of stack promoted objects`

eeckstein reviewed (commented):
This looks already pretty good.
I have some further comments - mostly nitpicks.
Most of the comments refer to things which are already not ideal on the C++ side. Let's try to be more "perfectionistic" in the swift implementation.

MaxDesiatov committed:
libswift: reimplement `Instruction` helpers in Swift

eeckstein commented on SwiftCompilerSources/Sources/SIL/Instruction.swift:
```diff
@@ -98,6 +98,19 @@ public class Instruction : ListNode, CustomStringConvertible, Hashable {
     }
   }
 
+  final public var mayReadRefCount: Bool {
```
I thought about this again. The reason why we have a `SILInstruction::mayReleaseOrReadRefCount()` is because every release can call a destructor with arbitrary side effects, including reading the reference count.
E.g. instructions like `strong_release` or `apply` can read the refcount.
But fortunately this check is not needed in the release-devirtualizer: it's not legal SIL to read the ref count after the final release. So it's sufficient to check for `mayRelease`.
You can just remove `mayReadRefCount` here and in the release-devirtualizer.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+
+private func stripRCIdentityPreservingInsts(_ value: Value) -> Value? {
+  switch value {
+  // First strip off RC identity preserving casts.
```
If you do a `guard let inst = value as? Instruction else { return }` before the switch, you can switch over `inst` which simplifies the code below:
``` 
case is UpcastInst, is UncheckedRefCastInst, ...
```

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+  // If we have a struct instruction with only one non-trivial stored field, the
+  // only reference count that can be modified is the non-trivial field. Return
+  // the non-trivial field.
+  case let si as StructInst:
```
you can combine StructInst and TupleInst:
```
case is StructInst, isTupleInst:
 return inst. ti.uniqueNonTrivialOperand
```

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+  // If we have an enum instruction with a payload, strip off the enum to
+  // expose the enum's payload.
+  case let ei as EnumInst where !ei.operands.isEmpty:
+    return ei.operands[0].value
```
`return ei.operand`

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+  // of a tuple, a retain_value on the tuple is equivalent to a retain_value on
+  // the extracted value.
+  case let tei as TupleExtractInst where tei.isEltOnlyNonTrivialElt:
+    return tei.operands[0].value
```
`return tei.operand`

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+  // the struct is equivalent to a ref count operation on the extracted
+  // member. Strip off the extract.
+  case let sei as StructExtractInst where sei.isFieldOnlyNonTrivialField:
+    return sei.operands[0].value
```
`return sei.operand`

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+    let function = self.function
+
+    // For each operand...
+    for op in operands.enumerated() {
```
no need for `enumerated()`

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+    var candidateElt: Value?
+    let function = self.function
+
+    // For each operand...
```
This comment is redundant - you can remove it.
Comments should describe things which are not obvious from the program code and not just do a 1:1 translation from swift syntax to English.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+
+    // For each operand...
+    for op in operands.enumerated() {
+      // If the operand is not trivial...
```
the same here: no need for this comment

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+    for op in operands.enumerated() {
+      // If the operand is not trivial...
+      if !op.type.isTrivial(in: function) {
+        // And we have not found a `candidateElt` yet, set index to `op` and continue.
```
and here: no need for this comment

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,214 @@
+}
+
+private extension TupleExtractInst {
+  var isEltOnlyNonTrivialElt: Bool {
```
For the use in the release-devirtualizer it would be simpler to add a `Type.countNonTrivialTupleElements` and bail if it's not 1.

MaxDesiatov committed:
libswift: implement `isFieldOnlyNonTrivialField`

MaxDesiatov committed:
libswift: simplify `ReleaseDevirtualizer.swift`

MaxDesiatov committed:
libswift: remove `Instruction.mayReadRefCount`

MaxDesiatov commented:
@swift-ci please smoke test

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,201 @@
+
+    let opType = operand.type
+
+    return opType.tupleElements.filter { 
```
Unfortunately `filter` is pretty bad for performance because it creates a new Array (involving memory allocation, etc.).
You can do `lazy.filter`, but then you cannot use `count`.
Or you do a good old fashioned `for ... in ...` loop (which would have the benefit that you abort as soon as the count exceeds one.

eeckstein commented on SwiftCompilerSources/Sources/Optimizer/FunctionPasses/ReleaseDevirtualizer.swift:
```diff
@@ -0,0 +1,201 @@
+
+    let structType = operand.type
+
+    return structType.getStructFields(in: function).filter { 
```
the same here

eeckstein reviewed (commented):
Very nice!
Still one small comment.

MaxDesiatov commented:
@swift-ci please smoke test

eeckstein reviewed (approved):
lgtm, thanks!

MaxDesiatov commented:
@swift-ci please smoke test

MaxDesiatov commented:
@swift-ci please test

MaxDesiatov commented:
C++ interop std_vector test failure is unrelated

MaxDesiatov commented:
@eeckstein thanks for your detailed review feedback and guidance!
